string input_msg
------
string output_res